rflib/chipcon_nic.py

```diff
@@ -763,7 +763,7 @@ def setMdmChanBW(self, bw, mhz=24, radiocfg=None):
         if chanbw_e is None:
             raise Exception("ChanBW does not translate into acceptable parameters.  Should you be changing this?")
 
-        bw = 1000.0*mhz / (8.0*(4+chanbw_m) * pow(2,chanbw_e))
```

Comment:
i'm not following, can you justify this change?  point me at a manual page or something?

Reply:
def setMdmChanBW(self, bw, mhz=24, radiocfg=None):
        
        if radiocfg==None:
            self.getRadioConfig()
            radiocfg = self.radiocfg

        chanbw_e = None
        chanbw_m = None
        for e in range(4):
            m = int(((old_div(mhz*1000000.0, (bw *pow(2,e) * 8.0 ))) - 4) + .5)        # rounded evenly
            if m < 4:
                chanbw_e = e
                chanbw_m = m
                break
        if chanbw_e is None:
            raise Exception("ChanBW does not translate into acceptable parameters.  Should you be changing this?")

        bw = 1000.0*mhz / (8.0*(4+chanbw_m) * pow(2,chanbw_e))
        #print "chanbw_e: %x   chanbw_m: %x   chanbw: %f kHz" % (e, m, bw)

        radiocfg.mdmcfg4 &= ~(MDMCFG4_CHANBW_E | MDMCFG4_CHANBW_M)
        radiocfg.mdmcfg4 |= ((chanbw_e<<6) | (chanbw_m<<4))
        self.setRFRegister(MDMCFG4, (radiocfg.mdmcfg4))

        # from http://www.cs.jhu.edu/~carlson/download/datasheets/ask_ook_settings.pdf
        if bw > 102e3:
            self.setRFRegister(FREND1, 0xb6)
        else:
            self.setRFRegister(FREND1, 0x56)

        if bw > 325e3:
            self.setRFRegister(TEST2, 0x88)
            self.setRFRegister(TEST1, 0x31)
        else:
            self.setRFRegister(TEST2, 0x81)
            self.setRFRegister(TEST1, 0x35)
`


The function with the issue as a whole without the comments. The issue is thus: with bw = 1000.0*mhz / (8.0*(4+chanbw_m) * pow(2,chanbw_e)) the two following if statements:

`if bw > 102e3:
            self.setRFRegister(FREND1, 0xb6)`
and 
`if bw > 325e3:
            self.setRFRegister(TEST2, 0x88)
            self.setRFRegister(TEST1, 0x31)`

Are NEVER true.  

You can see this via the following:
chanbw_e  must be 0, 1, 2, or 3 (for e in range(4):)
chanbw_m must also be 0, 1, 2, or 3 (if m < 4:)

Given chanbw_e = 0 chanbw_m =0 and your formula:
bw = 1000.0*mhz / (8.0*(4+chanbw_m) * pow(2,chanbw_e))
bw = 750.0 

Given chanbw_e = 1 chanbw_m =1 and your formula:
bw = 1000.0*mhz / (8.0*(4+chanbw_m) * pow(2,chanbw_e))
bw = 300.0

Given chanbw_e = 2 chanbw_m =2 and your formula:
bw = 1000.0*mhz / (8.0*(4+chanbw_m) * pow(2,chanbw_e))
bw = 125.0

Given chanbw_e = 3 chanbw_m =3 and your formula:
bw = 1000.0*mhz / (8.0*(4+chanbw_m) * pow(2,chanbw_e))
bw = 53.571

(Yes I realize you can have 0 and 3 and other combos, but it is mathematically impossible for those if statements to be true with the bw formula only using 1000)

I assumed it was a typo on the bw formula given that exact same formula (but with the extra 0's added like my "fix") can be found in line 793 of chipcon_nic.py in the function getMdmChanBW, and the if statements *can* be true with 1000000.0. So you are doing it on one but not the other, both of which are ChanBW. Seen here:

`bw = 1000000.0*mhz / (8.0*(4+chanbw_m) * pow(2,chanbw_e))`

Reply:
thank you.  that was even more than i expected, and very understandable and convincing.  i also checked the manual, and realized that you're absolutely right.